### ISSUE TITLE
Add `Transforms.rewrapNode` method

### ISSUE THREAD

**Issue by thesunny** [ianstormtaylor/slate#4173]
**Problem**
It's common to want to take a node and replace it with another node while keeping all its children. A common use case scenario is converting a `paragraph` to a `heading` or a `list-item`. A typical way of doing this might be to change the properties but you lose type safety when you do that. For example, if you convert a `type: "paragraph"` to a `type: "heading"` but the heading type requires a `"level"` property, you've now created an invalid node.

**Solution**
Add a method named `Transforms.rewrapNode` that takes a container `Node`, removes it, inserts a new container `Node` but with the same children.

**Alternatives**
I think many people might just replace the node `type` and write logic to fix up the properties. But this is unsafe. Or they write the function. It's short but perhaps non-trivial for a new Slate user.

**Context**
I've implemented this already so I just want to get an okay from @ianstormtaylor on the addition the API. I can PR and merge it as the code exists.

**Comment by ianstormtaylor** [comment]
This brings up a bigger question, which is how to treat typings for node-based transforms. 

Because even in the case where you set all the properties of a node, the children could still be invalid for the new node type. It makes me want to learn towards having the node transforms not be typed at all, but I’m unsure.

**Comment by thesunny** [comment]
That's a great question!

My first instinct is to punt it for now. I feel so much better having type safety around Nodes and it might be nice to feel our way around that first to see where it feels like it's not doing enough.

I feel like I've come out of the wild west into a safety zone with the new types but maybe in a few months I'll feel like it's still too loose.

I've started and stopped writing a response several times now related to the logistics. My current train of thought which I'm not convinced is correct is that it may be possible to add type safety to some methods. For example, let's say we wanted to add a form of type safety to `Transforms.insertNodes`.

We could allow a generic that specifies the parent. The developer would still be responsible for passing that in, but we can allow users to add additional type safety if they want:

```typescript
const parentNode = Node.get(editor, path)

// this would valididate childNodes to be children of the typeof parentNode
Transforms.insertNodes<typeof parentNode>(editor, childNodes, {at: [...path, 0]})

// this would throw a TypeScript error because there is no type safety
Transforms.insertNodes(editor, childNode, {at: [...path, 0]})

// force the use of `any` to get around the TypeScript error
Transforms.insertNodes<any>(editor, childNode, {at: [...path, 0]})

// force the use of `any` to get around the TypeScript error (also valid `any`)
Transforms.insertNodes(editor, childNode as any, {at: [...path, 0]})

// And if we're confident that the parent is a table
Transforms.insertNodes<TableElement>(editor, childNode, {at: [...path, 0]})
```

One issue is that potentially we're dealing with a lot of `any` because in the process of normalization, we actually expect to create invalid state on our way to a valid state. For example, we may liftNodes with the expectation that it may be invalid until the next normalization pass.

**Comment by ianstormtaylor** [comment]
Yeah that kind of normalization issue was why I hesitated as well. Although thinking about it more, it might be alright to just allow for any valid editor node, force them to use `any` when it's invalid, and not handle parent-to-direct-children out of the box.

**Comment by thesunny** [comment]
As a follow up, that code I wrote above won't work. At least the `typeof parentNode` won't work because it would always be `Node`.

I think a reasonable option would be to allow a generic for the Node so that at least you can check you aren't doing something stupid.

```
// this would throw a type error because heading requires a `level`
Transforms.insertNodes<HeadingElement>(editor, {type: "heading:, children: [{text: '...'}]})

// We could have it default to Node and it would also catch the heading issue
Transforms.insertNodes(editor, {type: "heading", children: [{text: '...']})

// But wouldn't catch this because we passed the wrong type of Node (paragraph) but wanted a heading
Transforms.insertNodes(editor, {type: "paragraph", children: [{text: '...']})
```

One thing that would cause friction is to force the generic which I'm sure some won't like the extra keystrokes. On the other hand, it would force users to be intentional about their types which I think would decrease bugs. For example, if I was forced to provide the generic here:

```
Transforms.insertNodes<HeadingElement|TableCellElement>(...)
```

I would, by nature of writing the above generic, immediately get the sense that I was doing something wrong (why are `HeadingElement` and `TableCellElement` together when they have different parents?).

Even forcing me to enter a generic that is a very permissive `Node` (which is pretty close to `any`) would probably indicate to me that I'm doing something dangerous. My thought would be, why is `Transforms.insertNodes` forcing me to tell it what I want to insert? Oh, wait, I can't insert any `Node`, it has to be a `Block` node so I better check that I'm not providing an Inline. I'd be forced to consider it and instinctually I would want it to be as narrow as possible.
